menshikov/l3/1/reverse.c

```diff
@@ -0,0 +1,344 @@
+
+    char delimiter_symbol = '/';
+    merged_file_path = strncpy(merged_file_path, dir_path, dir_path_length);
+    merged_file_path[dir_path_length] = delimiter_symbol;
+
+    merged_file_path = strncat(merged_file_path, file_name, file_name_length);
+    return merged_file_path;
+}
+
+void backward_copying_files(FILE* input_file, FILE* output_file) {
+    int seeking_status = fseek(input_file, 0, SEEK_END);
+    if (seeking_status == SOMETHING_WENT_WRONG) {
+        perror("Error during fseek");
+        return;
+    }
+    long file_position = ftell(input_file);
+    if (file_position == SOMETHING_WENT_WRONG) {
+        perror("Error during ftell");
+        return;
+    }
+
+    char reading_buffer[BUFFER_SIZE];
+    long current_offset_from_end = 0;
+    while (file_position > 0) {
+        long count_of_reading_symbols = BUFFER_SIZE;
+        if (count_of_reading_symbols > file_position) {
+            count_of_reading_symbols = file_position;
+        }
+
+        current_offset_from_end += count_of_reading_symbols;
+        seeking_status = fseek(input_file, -current_offset_from_end, SEEK_END);
+        if (seeking_status == SOMETHING_WENT_WRONG) {
+            perror("Error during fseek");
+            return;
+        }
+
+        size_t return_read_code = fread(reading_buffer, sizeof(char), count_of_reading_symbols, input_file);
+        if (return_read_code < count_of_reading_symbols || (ferror(input_file) != OK)) {
+            fprintf(stderr, "Error during fread\n");
+            return;
+        }
+
+        for (long i = count_of_reading_symbols - 1; i >= 0; --i) {
+            fprintf(output_file, "%c", reading_buffer[i]);
+        }
+        file_position -= count_of_reading_symbols;
+    }
+}
+
+enum statuses_of_copy copy_file_content(char* dir_path, char* file_name, char* dir_path_with_reversed_name) {
+    char* file_path = prepare_file_path(dir_path, file_name);
+    if (file_path == NULL) {
+        return MEMORY_ALLOCATION_ERROR;
+    }
+
+    size_t file_name_length = strlen(file_name);
+    char reversed_file_name[file_name_length];
+    reverse_string(file_name, reversed_file_name);
+    char* reversed_file_path = prepare_file_path(dir_path_with_reversed_name, reversed_file_name);
+    if (reversed_file_path == NULL) {
+        free(file_path);
+        return MEMORY_ALLOCATION_ERROR;
+    }
+
+    FILE* input_file = fopen(file_path, "rb");
+    if (input_file == NULL) {
+        perror("Error during fopen");
+        free(file_path);
+        free(reversed_file_path);
+        return CANNOT_OPEN_FILE;
+    }
+    free(file_path);
+
+    FILE* output_file = fopen(reversed_file_path, "wb");
+    if (output_file == NULL) {
+        perror("Error during fopen");
+        free(reversed_file_path);
+        fclose(input_file);
+        return CANNOT_OPEN_FILE;
+    }
+    free(reversed_file_path);
+
+    backward_copying_files(input_file, output_file);
+    fclose(input_file);
```

Comment:
>   Dealing with error returns from close()
A  careful  programmer will check the return value of close(), since it is quite possible that errors on a previous write(2) operation are reported only on the final close() that releases the open file description.  Failing to check the return value when closing a file may lead to silent loss of data.  This can especially be observed with NFS and with disk quota.

Reply:
В одном файле должны быть одинаковые отступы, не смешивая пробелы и табуляцию
![image](https://github.com/NSU-FIT-Operating-Systems/21202-operating-systems-part1/assets/17474821/f248fdc9-4a12-42d8-89c8-fdb76916d1ff)

Reply:
В си есть так называемый short-circuit evaluation
Если первый файл не получилось закрыть, хотелось бы попробовать закрыть второй файл
> Unlike the bitwise | operator, the || operator guarantees left-to-right evaluation. If the first operand compares unequal to 0, the second operand is not evaluated.